src/callbacks_step/amr_dg1d.jl

```diff
@@ -3,255 +3,255 @@
-
-    resize!(u_ode, nvariables(equations) * nnodes(dg)^ndims(mesh) * nelements(dg, cache))
-    u = wrap_array(u_ode, mesh, equations, dg, cache)
-
-    # Loop over all elements in old container and either copy them or refine them
-    element_id = 1
-    for old_element_id in 1:old_n_elements
-      if needs_refinement[old_element_id]
-        # Refine element and store solution directly in new data structure
-        refine_element!(u, element_id, old_u, old_element_id,
-                        adaptor, equations, dg)
-        element_id += 2^ndims(mesh)
-      else
-        # Copy old element data to new element container
-        @views u[:, .., element_id] .= old_u[:, .., old_element_id]
-        element_id += 1
-      end
+    # Return early if there is nothing to do
+    if isempty(elements_to_refine)
+        return
+    end
+
+    # Determine for each existing element whether it needs to be refined
+    needs_refinement = falses(nelements(dg, cache))
+    needs_refinement[elements_to_refine] .= true
+
+    # Retain current solution data
+    old_n_elements = nelements(dg, cache)
+    old_u_ode = copy(u_ode)
+    GC.@preserve old_u_ode begin # OBS! If we don't GC.@preserve old_u_ode, it might be GC'ed
+        old_u = wrap_array(old_u_ode, mesh, equations, dg, cache)
+
+        # Get new list of leaf cells
+        leaf_cell_ids = local_leaf_cells(mesh.tree)
+
+        # re-initialize elements container
+        @unpack elements = cache
+        resize!(elements, length(leaf_cell_ids))
+        init_elements!(elements, leaf_cell_ids, mesh, dg.basis)
+        @assert nelements(dg, cache) > old_n_elements
+
+        resize!(u_ode,
+                nvariables(equations) * nnodes(dg)^ndims(mesh) * nelements(dg, cache))
+        u = wrap_array(u_ode, mesh, equations, dg, cache)
+
+        # Loop over all elements in old container and either copy them or refine them
+        element_id = 1
+        for old_element_id in 1:old_n_elements
+            if needs_refinement[old_element_id]
+                # Refine element and store solution directly in new data structure
+                refine_element!(u, element_id, old_u, old_element_id,
+                                adaptor, equations, dg)
+                element_id += 2^ndims(mesh)
+            else
+                # Copy old element data to new element container
+                @views u[:, .., element_id] .= old_u[:, .., old_element_id]
+                element_id += 1
+            end
+        end
+        # If everything is correct, we should have processed all elements.
+        # Depending on whether the last element processed above had to be refined or not,
+        # the counter `element_id` can have two different values at the end.
+        @assert element_id ==
+                nelements(dg, cache) +
+                1||element_id == nelements(dg, cache) + 2^ndims(mesh) "element_id = $element_id, nelements(dg, cache) = $(nelements(dg, cache))"
```

Comment:
:cry:

Reply:
Macros are super problematic for JuliaFormatter, unfortunately :-/

This is unfixable except by disabling formatting here